### PR TITLE
workaround to support custom extensions that use standard prefixes

### PR DESCRIPTION
**Pull request by arromanoff**
RISC-V ISA states (21.1):
> A standard-compatible global encoding can also use standard prefixes for non-standard extensions if the associated standard extensions are not included in the global encoding."

Currently all the instructions (either from standard or custom extensions) are all being inserted into a single std::vector which is then being sorted. An instruction matching process performs linear search on that vector. The problem is that when a custom extension uses the same opcode as standard one (i.e. match and mask are equal to the standard counterparts) it is undefined which instruction will be picked. That is because in std::sort "The order of equal elements is not guaranteed to be preserved". That being said it is impossible to define custom extension (via customext) that would use the prefix of a disabled standard extension.

In this change I separate custom and standard extensions in two separate std::vector's. By default we report an error if they have common elements (There're an additional processor_t constructor's argument that skips this check). If this error is disabled during instruction matching we first trying to find it among custom instructions. If it has been found the search is stopped and custom instruction is executed, otherwise we look for it among standard instructions. Overall this change does not completely fix the problem but at least makes it possible to use the feature of RISC-V ISA.

### CODE REVIEW

**Comment by arromanoff** [comment]
@aswaterman , @timsifive could you please take a look?

File: riscv/processor.cc [hunk 1]
```diff
@@ -61,7 +61,7 @@ processor_t::processor_t(const isa_parser_t *isa, const cfg_t *cfg,
 
   disassembler = new disassembler_t(isa);
   for (auto e : isa->get_extensions())
-    register_extension(find_extension(e.c_str())());
+    register_extension(find_extension(e.c_str())(), true /* is_custom */);
```
Comment (jerryz123) [inline riscv/processor.cc]: I think having separate `register_base_extension` and `register_custom_extension` is cleaner.

**Review by aswaterman — COMMENTED** [review]
My inclination is to assume the authors of custom extensions know what they're doing, and allow them to override standard behavior at their will.  In that case, we could avoid adding a new command-line argument.  What do you think @jerryz123?

**Comment by jerryz123** [comment]
> In that case, we could avoid adding a new command-line argument.

I agree with this. A warning might be appropriate, but there is an argument that spike should not dump anything to console that is not from the target binary when executed without the debug/logging flags.

**Comment by aswaterman** [comment]
For my part, I'd even forgo the warning.

**Comment by arromanoff** [comment]
> My inclination is to assume the authors of custom extensions know what they're doing, and allow them to override standard behavior at their will. In that case, we could avoid adding a new command-line argument. 

Are you suggesting to move this "overrides standard extension" leverage into `extension_t` so that extension authors can express their intents? I like that idea. There's a problem that behavior will still be undefined if user overrides standard opcode unintentionally but it is probably ok for now.

**Comment by aswaterman** [comment]
I was more straightforwardly suggesting that we get rid of the command-line option and make Spike act as though `--extension-intersection-allowed=true` had been passed.

We can presumably use a `stable_sort` instead so that the custom extensions are deterministically picked over the standard ones (which might involve reversing the vector first or something like that).

**Comment by arromanoff** [comment]
`reverse` + `stable_sort` might not be the best idea since we may have several such extensions and have to perform this sorting (reversing) procedure several times.

File: riscv/processor.cc (hunk 1 above)
Comment (arromanoff) [reply riscv/processor.cc]: Agreed. Changed.

**Review by aswaterman — APPROVED** [review]
OK with me, but I'll give @jerryz123 a chance to weigh in

**Review by jerryz123 — APPROVED** [review]
Thanks for adding the testing

**Comment by aswaterman** [comment]
I forgot I need to update the CI test tarball to make the tests pass in CI.  However, I couldn't get the new tests to pass locally.  I get:

`test-customext: /scratch/waterman/riscv-isa-sim/ci-tests/../fesvr/elfloader.cc:45: std::map<std::__cxx11::basic_string<char>, long unsigned int> load_elf(const char*, memif_t*, reg_t*, unsigned int): Assertion IS_ELF_RISCV(*eh64) || IS_ELF_EM_NONE(*eh64) failed.`

Can you get them working so I can build the tarball and upload it?  From the root directory, run `ci-tests/build-spike` then `ci-tests/test-spike`.  (You need to run from the root directory to match the CI behavior.)

**Comment by arromanoff** [comment]
Problem solved. Now tests are working again (at least locally).

**Comment by jerryz123** [comment]
@aswaterman I guess you've updated the tarball if the CI is passing? This can be merged now?

**Comment by aswaterman** [comment]
Yeah, I actually uploaded the tarball yesterday when the tests were still failing.  I guess the problem was not with the tarball contents.  Out of caution, I regenerated it and made sure the tests are still passing locally (they are) and uploaded it again.

**Pull request merged**